akd_core/src/utils.rs

```diff
@@ -33,37 +42,140 @@ pub fn get_marker_version_log2(version: u64) -> u64 {
+    if MARKER_VERSION_SKIPLIST[skiplist_past_index] != start_version {
+        past_marker_versions.push(MARKER_VERSION_SKIPLIST[skiplist_past_index]);
+    }
+    let start_version_log2 = 1 << get_marker_version_log2(start_version);
+    if start_version_log2 != start_version
+        && (past_marker_versions.is_empty()
+            || start_version_log2 != past_marker_versions[past_marker_versions.len() - 1])
+    {
+        past_marker_versions.push(start_version_log2);
+    }
+
+    let start_version_length = get_bit_length(start_version);
+    for i in (0..start_version_length).rev() {
+        let shift = 1 << i;
+        // Check if the bit of start_version at position i is 1
+        if start_version & shift != 0 {
+            let shift_mask = (shift - 1) | shift;
+            let past = start_version & !shift_mask;
+            if past != 0
+                && (past_marker_versions.is_empty()
+                    || past != past_marker_versions[past_marker_versions.len() - 1])
+            {
+                past_marker_versions.push(past);
+            }
+        }
     }
 
     // Compute future marker versions
+    let mut future_marker_versions: Vec<u64> = Vec::new();
+
+    let end_version_length = get_bit_length(end_version);
+    let mut future: u64 = end_version;
+    for i in 0..end_version_length {
+        let shift = 1 << i;
+        // Check if the bit of end_version at position i is 0
+        if end_version & shift == 0 {
+            future |= shift;
+            future &= !(shift - 1);
+            if future <= epoch {
+                future_marker_versions.push(future);
+            }
+        }
+    }
+
+    let endv_index: usize = find_max_index_in_array(end_version, &MARKER_VERSION_SKIPLIST);
+    let epoch_index: usize = find_max_index_in_array(epoch, &MARKER_VERSION_SKIPLIST);
+    let skiplist_slice = &MARKER_VERSION_SKIPLIST[endv_index + 1_usize..epoch_index + 1_usize];
+
     let next_marker_log2 = get_marker_version_log2(end_version) + 1;
     let final_marker_log2 = get_marker_version_log2(epoch);
-    let mut future_marker_versions: Vec<u64> =
-        ((end_version + 1)..(1 << next_marker_log2)).collect();
     for i in next_marker_log2..(final_marker_log2 + 1) {
+        let val = 1 << i;
+        if !skiplist_slice.is_empty() && val >= skiplist_slice[0] {
+            // Don't need to add any more powers of 2, can just append the skiplist slice
+            break;
+        }
         future_marker_versions.push(1 << i);
     }
+    future_marker_versions.extend_from_slice(skiplist_slice);
 
     (past_marker_versions, future_marker_versions)
 }
 
+// Given an input u64 and a sorted array of u64s, find the largest index for which
+// the corresponding array element is less than the input.
+//
+// Note that if the input is less than the smallest element of the array, or the
+// array is of length 0, then this function will panic.
+fn find_max_index_in_array(input: u64, arr: &[u64]) -> usize {
```

Comment:
To ensure I'm not missing anything, the arr in this case is usually a small sorted array; right? I ask because we can technically do a binary search on the arr (since it's sorted) to find the largest index. However, if the array is small, I imagine the linear search may actually be faster in practice. I don't think this matters too much in the grand scheme of things, but figured I'd mention it.

Reply:
Good call! Hmm... so I changed this function to actually be called `find_max_index_in_skiplist()` since we only ever call it on MARKER_VERSION_SKIPLIST. And I added a comment that we could optimize this to be a binary search instead of a linear search in the future, but that at the moment it doesn't make too much of a difference since the array is only 7 elements long.